src/rgw/driver/sfs/types.cc

```diff
@@ -18,37 +18,195 @@
 #include "rgw/rgw_sal_sfs.h"
 #include "rgw/driver/sfs/types.h"
 #include "rgw/driver/sfs/object_state.h"
+#include "sqlite_objects.h"
 
 namespace rgw::sal::sfs {
 
-std::filesystem::path Object::get_storage_path() const {
-  return path.to_path() / std::to_string(version_id);
+Object::Object(const std::string& _name, const uuid_d& _uuid)
+    : name(_name), path(_uuid), deleted(false) {
+}
+
+Object* Object::create_for_immediate_deletion(
+    const sqlite::DBOPObjectInfo& object) {
+  Object* result = new Object(object.name, object.uuid);
+  result->deleted = true;
+  return result;
+}
+
+Object* Object::create_for_query(const std::string& name, const uuid_d& uuid,
+                                 bool deleted, uint version_id) {
+  Object* result = new Object(name, uuid);
+  result->deleted = deleted;
+  result->version_id = version_id;
+  return result;
+}
+
+Object* Object::create_for_testing(const std::string& name) {
+  Object* result = new Object(name, UUIDPath::create().get_uuid());
+  return result;
+}
+
+Object* Object::create_from_obj_key(const rgw_obj_key& key) {
+  Object* result = new Object(key);
+  return result;
+}
+
+Object* Object::create_for_multipart(const std::string& name) {
+  Object* result = new Object(name, UUIDPath::create().get_uuid());
+  return result;
```

Comment:
Are all these really needed?

I see them as constructor helpers.

`create_for_testing` and `create_for_multipart` are the same.  (at least for now, it they diverge at some point happy to have 2 different methods)

What about a single `create_from_name`? (or even a constructor with a single parameter which is the name)

And the `create_for_immediate_deletion` seems a bit weird to me. 
If the object is immediately deleted I guess we don't need to set the `deleted` attribute to false.
I've tested this already removing the `result->deleted = true;` statement and as expected the `GC` tests pass.

I think calling the already existing `Object(const std::string& _name, const uuid_d& _uuid)` would be enough and we can save having that method around. (less code to be maintained) 

I dunno, maybe it is a personal thing... but... to me...
`create_from_obj_key` would be just:
```c++
Object(const rgw_obj_key & key)
```

The one with only the name would be: 
```c++
Object(const std::string& name)
```

and I would maybe just keep the helper for the query thing.

Reply:
The named constructors try to make the more clear what Object is created for. We sometimes use it to retrieve data, sometimes update data, sometimes create a new object. I'd like to remove some of the use cases from Object over time. The named constructors are a first step in that direction. They do feel like overkill right now, but will make it easier to remove dependencies from Object IMHO.